### PR TITLE
mf:// fixes

### PR DESCRIPTION
**Pull request by guidocella**

### CODE REVIEW

**Comment by avih** [comment]
> [demux_mf: don't intrepret % sequences in filenames](https://github.com/mpv-player/mpv/pull/13382/commits/2f52e8ff771fc573428dc79f25b4f4e838bdabe8)
>
> This feature is totally broken With `foo%.jpg`, `mpv mf://foo%%.jpg`

This example doesn't work because it doesn't have `%d`, so the frame number won't advance, so it's rejected as invalid format for the `mf://` demuxer. It's explained at [this comment](https://github.com/mpv-player/mpv/commit/d0c530919d#diff-0d847e49633638550fa1ae8e38f2ec3e503acd10ce18ae615d5aced437840a82R177) of the source code.

If you do have `%d`, then `%%` is processed correctly, for instance, generate 10 images with names `image%-<NUMBER>-%.jpg`:
```
ffmpeg -f lavfi -i testsrc=duration=1:rate=10 image%%-%d-%%.jpg
```

Plays fine in mpv:
```
mpv -v mf://image%%-%d-%%.jpg
```

> [it also doesn't work if there is] `foo11.jpg`, `mpv mf://foo%d.jpg`

It would work if you also have `foo1.jpeg`...`foo10.jpg`.

Because this reads numbered files until accumulating 5 missing numbers, then it stops looking for more. So it tries 1-5 (or 0-4, I don't recall), and then stops looking for further numbers, like 6 or 7 or 11 or 100001.

Test case:
generate 30 files image1.jpg ... image30.jpg:
```
ffmpeg -f lavfi -i testsrc=duration=1:rate=30 image%d.jpg
```

Play them in mpv:
```
mpv -v mf://image%d.jpg
```

And you'll notice that it loads all the images 1...30, including 11.

> and even mpv mf://foo%.2d.jpg don't work.

I'm not sure if this is still about the specific file `foo11.jpg`, or a general complaint that `%.2d` doesn't work.

If it's about the specific 11 file, then this is explained above, where there are more than 5 missing numbers before 11.

If this is generally about `%.2d`, then it does work for me:

Generate 30 files `image01.jpg`, `image02.jpg`...`image30.jpg`:
```
ffmpeg -f lavfi -i testsrc=duration=1:rate=30 image%02d.jpg
```

Play in mpv:
```
mpv -v mf://image%.2d.jpg
```
or
```
mpv -v mf://image%02d.jpg
```

(`%.2d` and `%02d` are identical in this case)

> No one noticed that this feature doesn't work for almost 3 years, since
https://github.com/guidocella/mpv/commit/d0c530919d8cd4d7a774e38ab064e0fabdae34e6 changed it, meaning that not a single user uses it.

Well, either that, or that it actually did work for people. I'm expecting the most common use case is either straight sequential numbers, like `foo1.bmp`, `foo2.bmp`, ...`foo42.bmp`..., or in some fixed width with `0` prefixes, like `foo0001.bmp`...`foo0042.bmp`...`foo1234.bmp`... .

And both these forms work fine, as well as all the cases you described as not working.

So this could be another reason why no one complained.

> Actually, mf:// has been completely broken for 3 months since https://github.com/guidocella/mpv/commit/ef11d31c3acfd71307a94e44ad164a4861287675 and no one noticed that either

Good catch. I'd hope that it can be fixed.

As for why no one noticed it, well, it's 3 months. Many mpv users don't update daily or even monthly. Heck, Debian, Fedora, and Ubuntu users are still few releases back, and that's not a small bunch of mpv users.

> Instead of fixing a feature that is undocumented, no one uses, and caused a security release, just remove it.

I still don't see what needs fixing, because it does seem to work as designed and generally, I'd presume, as expected.

The security issue was fixed very quicky, with 4 pairs of eyeballs reviewing this patch, so it's probably fair to assume this code is now less likely to have security issues than most other places at the mpv code.

I don't know who uses them, but the fact that it matches an ffmpeg feature which generates numbered frames (in the sense that mpv can play these numbered frames, with same format specified as the ffmpeg output format specifier), suggests to me that it can have use cases similar to others where users choose mpv to play things instead of ffplay. But that's a guess.

The only valid rant is that it's undocumented, so feel free to add docs.

FWIW, personally I don't think it's a bad feature to have in mpv, especially when it matches a known ffmpeg output form.

**Comment by guidocella** [comment]
I understood how this works now and restored and documented the behavior. mf:// being broken is fixed by commit 1.

**Comment by avih** [comment]
It's probably better to document it as `printf-format-specifier` or `printf-format` instead of `printf_pattern`, because its official name is "format specifier" or `"format string". It's not a pattern.

It might be also worth documenting the format subset which is accepted (because we don't accept all valid printf formats which take one int argument), and that's `%[.][NUM]d`, whete NUM is one, two, or three decimal digits.

If you  feel like being more user friendly about why a format was rejected, then you can also mention at the rejection error message that exactly one format of the form `%[.][NUM]d` is expected.

File: video/out/vo.c [hunk 1]
```diff
@@ -1196,9 +1196,16 @@ void vo_seek_reset(struct vo *vo)
 bool vo_still_displaying(struct vo *vo)
 {
     struct vo_internal *in = vo->in;
```
Comment (sfan5) [inline video/out/vo.c:1198]: IMO this commit should keep the `Revert "..."` message and only mention mf:// in the description

File: DOCS/man/mpv.rst [hunk 2]
```diff
@@ -1200,9 +1200,17 @@ PROTOCOLS
 
     Digital TV via DVB. (Linux only.)
 
-``mf://[filemask|@listfile]`` ``--mf-...``
-
-    Play a series of images as video.
+``mf://[@listfile|filemask|glob|printf-format]`` ``--mf-...``
+
+    Play a series of images as a video. If the URL begins with ``@``, it is
+    interpreted as a sequence of paths separated by newlines. If the URL
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1206]: I don't think the *URL* is interpreted as that. Rather it reads a file at that path and interprets *its contents*.
No?

File: DOCS/man/mpv.rst [hunk 3]
```diff
@@ -1200,9 +1200,17 @@ PROTOCOLS
 
     Digital TV via DVB. (Linux only.)
 
-``mf://[filemask|@listfile]`` ``--mf-...``
-
-    Play a series of images as video.
+``mf://[@listfile|filemask|glob|printf-format]`` ``--mf-...``
+
+    Play a series of images as a video. If the URL begins with ``@``, it is
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1205]: would put a newline + empty line here

File: DOCS/man/mpv.rst [hunk 4]
```diff
@@ -1200,9 +1200,17 @@ PROTOCOLS
...
+    interpreted as a sequence of paths separated by newlines. If the URL
+    contains ``,``, it is interpreted as a sequence of paths separated by ``,``.
+    If the URL doesn't contain ``%`` and if on POSIX platforms, the URL is
+    interpreted as a glob, and ``*`` is automatically appended if it was not
+    specified. Otherwise, the printf sequences ``%[.][NUM]d``, where ``NUM`` is
+    one, two, or three decimal digits, and ``%%`` and are interpreted. For
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1211]: I don't think there's a need to be so exact but :shrug:

File: DOCS/man/mpv.rst [hunk 5]
```diff
@@ -1200,9 +1200,17 @@ PROTOCOLS
...
+    example, ``mf://image-%d.jpg`` plays files like ``image-1.jpg``,
+    ``image-2.jpg`` and ``image-10.jpg``.
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1213]: it doesn't actually
there are more than 5 missing between 2 and 10 so it would only play the first two

**Comment by avih** [comment]
> No one noticed that this feature doesn't work for almost 3 years, since d0c530919d8cd4d7a774e38ab064e0fabdae34e6 changed it

As we demonstrated clearly, it does work for normal use cases of sequential numbered images, so it wouldn't be true to say it doesn't work.

It's true, though, that your specific examples don't work.

However, for the record, these specific examples never worked also before this commit, so you might as well have picked any random mpv commit as the reason for this, and I don't know why you picked this one without even checking that it actually changed the behavior with your examples.

It would have been trivial to revert it and check the behavior.

The reason your examples don't work, as explained above, is because they're outside the scope of what this feature was designed to handle, and this didn't change ever since it was added to mplayer in 2002.

The only things this commit does is explicit rejection of format specifiers which can't be used validly with this feature anyway, and avoiding buffer overflow in some rare cases with a maliciously crafted format specifier.

If you do notice any actual regressions with this commit, do tell.

**Comment by guidocella** [comment]
What? I removed that commit.

**Comment by avih** [comment]
> What? I removed that commit.

Yes, but beyond the fact that the examples were explained, I wanted to note explicitly that the behavior of these examples was not actually changed by this commit.

So that in the future you'd be more careful in your statements and commit messages.

File: DOCS/man/mpv.rst [hunk 6]
```diff
@@ -1200,10 +1200,19 @@ PROTOCOLS
 
     Digital TV via DVB. (Linux only.)
 
-``mf://[filemask|@listfile]`` ``--mf-...``
+``mf://[@listfile|filemask|glob|printf-format]`` ``--mf-...``
 
     Play a series of images as video.
 
+    If the path begins with ``@``, it is interpreted as a sequence of paths
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1207]: > If the path begins with `@`, it is interpreted as a sequence of paths separated by newlines

this is still incorrect, see previous comment

File: DOCS/man/mpv.rst [hunk 7]
```diff
@@ -1200,10 +1200,19 @@ PROTOCOLS
...
+    separated by newlines. If the path contains ``,``, it is interpreted as a
+    sequence of paths separated by ``,``. If the path does not contain ``%`` and
+    if on POSIX platforms, the path is interpreted as a glob, and ``*`` is
+    automatically appended if it was not specified. Otherwise, the printf
+    sequences ``%[.][NUM]d``, where ``NUM`` is one, two, or three decimal
+    digits, and ``%%`` and are interpreted. For example, ``mf://image-%d.jpg``
```
Comment (sfan5) [inline DOCS/man/mpv.rst:1213]: suggestion:
> For example, `mf://image-%d.jpg` plays files like `image-1.jpg`, `image-2.jpg` and `image-10.jpg`
> provided there are no big gaps in the files that exist.

File: DOCS/man/mpv.rst (hunk 6 above)
Comment (guidocella) [reply DOCS/man/mpv.rst]: The part of a URL after the scheme is called path, I don't know how else you want to call it? For now I specificied URL path to differentiate it from file paths it contains.

**Comment by kasper93** [comment]
Reverting https://github.com/mpv-player/mpv/commit/ef11d31c3acfd71307a94e44ad164a4861287675 will likely break https://github.com/mpv-player/mpv/issues/12575 could you check that before?

**Comment by guidocella** [comment]
Yeah it randomly freezes after seeks with the script in that issue.

File: DOCS/man/mpv.rst (hunk 6 above)
Comment (sfan5) [reply DOCS/man/mpv.rst]: No, not that part.

> **it is interpreted as a sequence of paths separated by newlines**

If I use `mf://@foo` mpv will go and open `foo` and treat every line in it as a path to be played.
That's not what the documentation says.

Comment (guidocella) [reply DOCS/man/mpv.rst]: Ah I've been testing with `mf://@<(print -l *.jpg)` and totally forgot it's supposed to accept a file.

**Comment by guidocella** [comment]
I removed the first commit for now as I have no idea how to fix both that crop issue and mf, hopefully kasper or someone else can figure it out later.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**